cmd/vault-subpath-proxy/kv_update_transport.go

```diff
@@ -96,9 +119,124 @@ func (k *kvUpdateTransport) RoundTrip(r *http.Request) (*http.Response, error) {
+
+	// Clear old entries associated with us
+	for _, existingEntry := range k.existingSecretKeysByVaultSecretName[path] {
+		k.existingSecretKeysByNamespaceName[existingEntry.name].Delete(existingEntry.key)
+	}
+	delete(k.existingSecretKeysByVaultSecretName, path)
+
+	// Create new entries
+	name := types.NamespacedName{Namespace: item[vault.SecretSyncTargetNamepaceKey], Name: item[vault.SecretSyncTargetNameKey]}
+	if k.existingSecretKeysByNamespaceName[name] == nil {
+		k.existingSecretKeysByNamespaceName[name] = sets.String{}
+	}
+	for key := range item {
+		if key == vault.SecretSyncTargetNamepaceKey || key == vault.SecretSyncTargetNameKey {
+			continue
+		}
+		k.existingSecretKeysByNamespaceName[name].Insert(key)
+		k.existingSecretKeysByVaultSecretName[path] = append(k.existingSecretKeysByVaultSecretName[path], namespacedNameKey{name: name, key: key})
+	}
+}
+
+func (k *kvUpdateTransport) validateKeysDontConflict(ctx context.Context, data map[string]string) (validationErrs []string, err error) {
+	if k.privilegedVaultClient == nil {
+		return nil, nil
+	}
+	k.existingSecretKeysByNamespaceNameLock.Lock()
+	defer k.existingSecretKeysByNamespaceNameLock.Unlock()
+	if k.existingSecretKeysByNamespaceName == nil {
+		k.existingSecretKeysByNamespaceName = map[types.NamespacedName]sets.String{}
+		k.existingSecretKeysByVaultSecretName = map[string][]namespacedNameKey{}
+		// Clear up the map if we had an error, to avoid caching an incomplete result
+		defer func() {
+			if err != nil {
+				k.existingSecretKeysByNamespaceName = nil
+			}
+		}()
+
+		everything, err := k.privilegedVaultClient.ListKVRecursively(k.kvMountPath)
+		if err != nil {
+			return nil, fmt.Errorf("listKVRecursively failed: %w", err)
+		}
+		// We fetch pretty much all data, use limited concurrency to not get oom killed
+		sema := semaphore.NewWeighted(10)
+		wg := &sync.WaitGroup{}
+		wg.Add(len(everything))
+
+		var fetchErrs []error
+		var fetchErrLock sync.Mutex
+
+		// Need a new lock to synchronize across the fetching goroutines
+		var existingSecretKeysByNamespaceNameWriteLock sync.Mutex
+		for _, path := range everything {
+			path := path
+			go func() {
+				defer wg.Done()
+				if err := sema.Acquire(ctx, 1); err != nil {
+					fetchErrLock.Lock()
+					fetchErrs = append(fetchErrs, err)
+					fetchErrLock.Unlock()
+					return
+				}
+				defer sema.Release(1)
+
+				item, err := k.privilegedVaultClient.GetKV(path)
+				if err != nil {
+					fetchErrLock.Lock()
+					fetchErrs = append(fetchErrs, err)
+					fetchErrLock.Unlock()
+					return
+				}
+
+				existingSecretKeysByNamespaceNameWriteLock.Lock()
+				defer existingSecretKeysByNamespaceNameWriteLock.Unlock()
+
+				name := types.NamespacedName{Namespace: item.Data[vault.SecretSyncTargetNamepaceKey], Name: item.Data[vault.SecretSyncTargetNameKey]}
+				if k.existingSecretKeysByNamespaceName[name] == nil {
+					k.existingSecretKeysByNamespaceName[name] = make(sets.String, len(item.Data)-2)
```

Comment:
We could `delete` the two keys we don't care about and remove the special case here and below.

Reply:
Done